system/include/SSE/xmmintrin.h

```diff
@@ -0,0 +1,709 @@
+#ifndef __emscripten_xmmintrin_h__
+#define __emscripten_xmmintrin_h__
+
+#include <wasm_simd128.h>
+
+#include <math.h>
+#include <string.h>
+
+#ifndef __SSE__
+#error "SSE instruction set not enabled"
+#endif
+
+// Emscripten SIMD support doesn't support MMX/float32x2/__m64.
+// However, we support loading and storing 2-vectors, so
+// treat "__m64 *" as "void *" for that purpose.
+typedef float __m64 __attribute__((__vector_size__(16), __aligned__(16)));
+typedef v128_t __m128;
+typedef v128_t __m128i;
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_set_ps(float __z, float __y, float __x, float __w)
+{
+  return wasm_f32x4_make(__w, __x, __y, __z);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_setr_ps(float __z, float __y, float __x, float __w)
+{
+  return wasm_f32x4_make(__z, __y, __x, __w);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_set_ss(float __w)
+{
+  return wasm_f32x4_make(__w, 0, 0, 0);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_set_ps1(float __w)
+{
+  return wasm_f32x4_splat(__w);
+}
+
+#define _mm_set1_ps _mm_set_ps1
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_setzero_ps(void)
+{
+  return wasm_f32x4_const(0.f, 0.f, 0.f, 0.f);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_load_ps(const float *__p)
+{
+  return *(__m128*)__p;
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadl_pi(__m128 __a, const void /*__m64*/ *__p)
+{
+  return __builtin_shufflevector(wasm_f32x4_make(((float*)__p)[0], ((float*)__p)[1], 0.f, 0.f), __a, 0, 1, 6, 7);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadh_pi(__m128 __a, const void /*__m64*/ *__p)
+{
+  return __builtin_shufflevector(__a, wasm_f32x4_make(((float*)__p)[0], ((float*)__p)[1], 0.f, 0.f), 0, 1, 4, 5);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadr_ps(const float *__p)
+{
+  __m128 __v = _mm_load_ps(__p);
+  return __builtin_shufflevector(__v, __v, 3, 2, 1, 0);
```

Comment:
These `__builtin_shufflevector`s are breaking the `v128_t` abstraction an relying on the fact that `v128_t` is defined to have four elements. Can you add a comment saying that these really should be wasm shuffle intrinsics as soon as the performance problem is figured out? Also, since we are already using types like `__f32x4` in this file, it would be good to use them here, too, to make the number of lanes explicit.

Reply:
Let me revisit this in next PR, I have conflicting changes to these lines in the next SSE2 PR.